kimchi/src/circuits/polynomials/foreign_mul/gate.rs

```diff
@@ -0,0 +1,528 @@
+                // Force this circuit gate on
+                GateType::ForeignMul0 => vec![F::one(), F::zero(), F::zero()],
+                GateType::ForeignMul1 => vec![F::zero(), F::one(), F::zero()],
+                GateType::ForeignMul2 => vec![F::zero(), F::zero(), F::one()],
+                _ => panic!("Invalid gate type {:?}", self.typ),
+            },
+        };
+        let evals = vec![eval(curr), eval(next)];
+
+        // Setup circuit constants
+        let constants = expr::Constants {
+            alpha: F::rand(rng),
+            beta: F::rand(rng),
+            gamma: F::rand(rng),
+            joint_combiner: F::rand(rng),
+            endo_coefficient: cs.endo,
+            mds: vec![],
+            foreign_modulus: cs.foreign_modulus.clone(),
+        };
+
+        let pt = F::rand(rng);
+
+        // Evaluate constraints
+        match linearized
+            .constant_term
+            .evaluate_(cs.domain.d1, pt, &evals, &constants)
+        {
+            Ok(x) => {
+                if x == F::zero() {
+                    Ok(())
+                } else {
+                    Err(format!("Invalid {:?} constraint", self.typ))
+                }
+            }
+            Err(_) => Err(format!("Failed to evaluate {:?} constraint", self.typ)),
+        }
+    }
+}
+
+/// Get vector of foreign field multiplication circuit gate types
+pub fn circuit_gates() -> Vec<GateType> {
+    vec![
+        GateType::ForeignMul0,
+        GateType::ForeignMul1,
+        GateType::ForeignMul2,
+    ]
+}
+
+/// Number of constraints for a given foreign field multiplication circuit gate type
+pub fn circuit_gate_constraint_count<F: FftField>(typ: GateType) -> u32 {
+    match typ {
+        GateType::ForeignMul0 => ForeignMul0::<F>::CONSTRAINTS,
+        GateType::ForeignMul1 => ForeignMul1::<F>::CONSTRAINTS,
+        GateType::ForeignMul2 => ForeignMul2::<F>::CONSTRAINTS,
+        _ => panic!("invalid gate type"),
+    }
+}
+
+/// Get combined constraints for a given foreign field multiplication circuit gate type
+pub fn circuit_gate_constraints<F: FftField>(typ: GateType, alphas: &Alphas<F>) -> E<F> {
+    match typ {
+        GateType::ForeignMul0 => ForeignMul0::combined_constraints(alphas),
+        GateType::ForeignMul1 => ForeignMul1::combined_constraints(alphas),
+        GateType::ForeignMul2 => ForeignMul2::combined_constraints(alphas),
+        _ => panic!("invalid gate type"),
+    }
+}
+
+/// Get the combined constraints for all foreign field multiplication circuit gate types
+pub fn combined_constraints<F: FftField>(alphas: &Alphas<F>) -> E<F> {
+    ForeignMul0::combined_constraints(alphas)
+        + ForeignMul1::combined_constraints(alphas)
+        + ForeignMul2::combined_constraints(alphas)
+}
+
+/// Foreign field multiplication CircuitGate selector polynomial
+#[serde_as]
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub struct SelectorPolynomial<F: FftField> {
```

Comment:
I think we should generalize this pattern to the other selector polynomials present in `ConstraintSystem`. Maybe create an issue for that if you don't want to do this here?

Reply:
Cool - nice.  Yes